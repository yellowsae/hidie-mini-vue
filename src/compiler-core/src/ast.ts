

// 导出枚举 
// 节点树的类型 

import { CREATE_ELEMENT_BLOCK } from "./runtimeHelpers"

export const enum NodeTypes {
  INTERPOLATION,
  STATEFUL_COMPONENT,
  ELEMENT,
  TEXT,
  ROOT,
  COMPOUND_EXPRESSION
}


export function createVNodeCall(context, tag, props, children) {
  context.helper(CREATE_ELEMENT_BLOCK)
  return {
    type: NodeTypes.ELEMENT,
    tag,
    props,
    children
  }
}
