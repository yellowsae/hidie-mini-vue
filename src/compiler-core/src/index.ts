

// compiler 编译器的出口 


// 直接导出 compiler 
export * from './compiler'




/**
 * 在什么时候调用 compiler 模块呢？
 * 
 * 在 runtime.core 中 把 setup 函数的返回值赋值后，会去调用 render 函数
 * - 因为最终的目的是用于写 template,  所以在调用 render 函数前 判断是否具有 template ， 然后把 template 编译为 render 函数 
 * 
 * 
 */

