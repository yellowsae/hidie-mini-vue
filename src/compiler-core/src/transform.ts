
/**
 * 实现：
 *  1. 使用深度优先遍历 节点树
 *  2. 深度优先遍历 搜索 需要修改的 text 
 *  3. 修改 text 的内容   
 */
export function transform(root, options = {}) { // root 根节点 
  // 传入options -> 用于修改数据的 
  // 使用创建一个全局上下文对象 -> 和实现 baseParse 的一样
  const context = createTransformerContext(root, options);



  // 1. 深度优先遍历
  // 传入全局上下文呢对象 context , 为下边使用
  traverseNode(root, context)

  // 2. 修改 content 的内容 


  // 创建一个 childrenRoot, 设置 root的根节点 
  // 在 codegen 使用 
  createRootCodegen(root)

}

function createRootCodegen(root) {
  // 设置 root.children[0]  根节点 -> 在 codegen 使用 
  root.codegenNode = root.children[0]
}

// 全局上下文对象 context
function createTransformerContext(root, options) {
  // 建立一个 context  {}
  const context = {
    // 存入 root 和 nodeTransformer
    root,
    nodeTransformer: options.nodeTransformer || []
  }
  return context
}


function traverseNode(node: any, context) {
  // 输出 
  // console.log(node)

  // 2. 修改 content 
  // 判断 node.type 的类型 是不是Text 
  // 程序变动点 
  // 不能这么直接写死 , 需要改为 动态设计 
  // if (node.type === NodeTypes.TEXT) {
  //   // 修改数据 
  //   node.content = node.content + "mini-vue"
  // }
  // 改为使用 nodeTransformer
  // 取出 nodeTransformer 
  const nodeTransformer = context.nodeTransformer
  // for 循环 遍历 
  for (let i = 0; i < nodeTransformer.length; i++) {
    // 取出 修改数据的函数 
    const transform = nodeTransformer[i]
    // 调用 
    // 并且把 node 节点传给 修改数据的函数
    transform(node)
  }
  // 修改 element


  // // 重构 抽离 处理 深度优先遍历的 children 逻辑 
  traverseChildren(node, context)
  // let children = node.children

  // // 当 node.children 有值时
  // // 递归执行深度优先遍历 - 程序稳定点
  // if (node.children) {
  //   // 循环 children
  //   for (let i = 0; i < children.length; i++) {
  //     // 取值
  //     const node = children[i]
  //     // 递归 
  //     traverseNode(node, context)
  //   }
  // }
}

function traverseChildren(node, context) {
  // 重构 抽离 处理 深度优先遍历的 children 逻辑 
  const children = node.children

  // 当 children 有值时
  // 递归执行深度优先遍历 - 程序稳定点
  if (children) {
    // 循环 children
    for (let i = 0; i < children.length; i++) {
      // 取值
      const node = children[i]
      // 递归 
      traverseNode(node, context)
    }
  }
}
