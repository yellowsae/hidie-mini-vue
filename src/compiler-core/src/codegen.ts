import { isString } from "../../shared"
import { NodeTypes } from "./ast"
import { CREATE_ELEMENT_BLOCK, helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers"



export function generate(ast) {

  // 实现根据 ast 生成的 render 函数

  // 重构 
  // - code 过多 
  // - += 过多
  // 实现一个全局上下文对象 保存 code
  const context = createCodegenContext()
  const { push } = context

  /**
   *   // 解析编译插值 
      // 1. 实现导入 Vue 模块函数 
      // import { toDisplayString as _toDisplayString } from "vue"
      const VueBinding = "vue"
      // const helpers = ['toDisplayString']  
      // helpers 应该存放到 transform 的逻辑上 , 绑定到 ast 树的根节点上 
      // 重命名 
      const aliasHelper = (s) => `${s} as _${s}`
      // 改为使用 ast.helpers 
      push(`import { ${ast.helpers.map(aliasHelper).join(', ')} } from "${VueBinding}"`)
      push('\n')
      // 伪实现
      // push(`import { toDisplayString as _toDisplayString } from "vue"`)  
      // 定义code -> 最终返回 
      // let code = ""
      // code += 'return '
      push('return ')
   */
  // 重构
  getFunctionPreamble(ast, context)


  // 函数名
  const functionName = "render"
  // 参数 
  const args = ['_ctx', '_cache']

  // 把 array 转为 string 
  const signature = args.join(', ')


  // 重构 
  push(`function ${functionName}(${signature}) {`)
  push(`return `)
  genNode(ast.codegenNode, context)
  push("}")
  // // 拼接函数
  // code += `function ${functionName}(${signature}) {`
  // // 函数体的内容 - 基于 ast 实现 
  // code += `return `
  // // const node = ast.children[0]
  // // 改为由 transform 提供的根节点 - 而不是 直接 取出 ast.children[0]
  // // 抽离 返回内容的逻辑 为一个函数 
  // code = genNode(code, ast)
  // // const node = ast.codegenNode
  // // code += `'${node.content}'`

  // // 尾部 }
  // code += '}'



  // 返回 code
  return {
    code: context.code
  }

  // 伪实现 
  // 返回 
  // return {
  //   code: `
  //   return function render(_ctx, _cache, $props, $setup, $data, $options) {
  //     return "hi 11"
  //   }    
  //   `
  // }
}



// 处理 编译器 头部导入模块的逻辑 
function getFunctionPreamble(ast, context) {
  const { push } = context
  const VueBinding = "Vue"
  // const helpers = ['toDisplayString']  
  // helpers 应该存放到 transform 的逻辑上 , 绑定到 ast 树的根节点上 
  // 重命名 
  const aliasHelper = (s) => `${helperMapName[s]}: _${helperMapName[s]}`
  // 改为使用 ast.helpers 

  // 判断 ast.helpers -> 因为 处理 text 不需要 导入 
  if (ast.helpers.length > 0) {
    push(`const { ${ast.helpers.map(aliasHelper).join(', ')} } = ${VueBinding}`)
  }
  push('\n')
  // 伪实现
  // push(`import { toDisplayString as _toDisplayString } from "vue"`)
  // 定义code -> 最终返回 
  // let code = ""
  // code += 'return '
  push('return ')
}




// 抽离 render 返回内容的逻辑 
// function genNode(code, ast) {
//   const node = ast.codegenNode
//   code += `'${node.content}'`
//   return code
// }
// 重构
function genNode(node, context) {

  // 使用 switch 根据类型 处理其他逻辑
  switch (node.type) {
    // 处理 text 
    case NodeTypes.TEXT:
      genText(node, context)
      break;

    // 插值 + Text {{message}}
    case NodeTypes.STATEFUL_COMPONENT:
      genExpression(node, context)
      break;
    // 处理 插值 
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context)
      break;

    // 处理 Element
    case NodeTypes.ELEMENT:
      genElement(node, context)
      break;


    // 处理中间层 compound 
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context)
      break;
  }

}

// 处理中间层 
function genCompoundExpression(node, context) {
  const { push } = context
  // compound.children -> 一个数组
  const children = node.children

  // for 循环 
  for (let i = 0; i < children.length; i++) {
    const child = children[i]
    // 判断当前节点是不是一个 string 
    // 也就是 判断是不是 + 号
    if (isString(child)) {
      push(child)
    } else {
      // 如果不是 + 号 ， 那就是 text | 插值 
      // 执行 genNode 递归调用就行
      genNode(child, context)
    }
  }
}


// 处理 render 函数返回值为 Element  
function genElement(node, context) {
  const { push, helper } = context
  // 取出children
  // 取出 props
  const { tag, children, props } = node
  // push(`${helper(CREATE_ELEMENT_BLOCK)}("${tag}")`)

  /**
   * 根据 element 的 children 生成 children 的逻辑
   * - 可以循环 children, 然后递归遍历 genNode 
   * - 出现问题
   *   - 拼接时候少了 + 号 和 _ctx
   * - 解决： 可以声明一个新的节点类型 ， 在element下面生成 (中间层类型 compound)， text 和 插值 的组合, 然后处理这个类型时候 添加上 + 号
   *    
   *   - 实现: 在 transformText 中处理 , 添加中间层 
   */
  // console.log('children', children)
  // 不需要下边的循环了， 改为 children[0] 中间层
  // const child = children[0]

  // 重构 element的 children  -> 让它赋值为 node.children


  // 实现三种联合解析 
  // push(`${helper(CREATE_ELEMENT_BLOCK)}("${tag}", null, 'hi, ' + _toDisplayString(_ctx.message))`)
  // push(`${helper(CREATE_ELEMENT_BLOCK)}("${tag}", ${props}, `)

  // 处理 tag | props | children 的逻辑
  push(`${helper(CREATE_ELEMENT_BLOCK)}(`)

  // 处理 props tag children 为 undefined 的情况
  // 如果是 undefined , 把它赋值为 null, 因为返回值是 [], 使用 genNodeList() 处理
  genNodeList(genNullable([tag, props, children]), context)


  // genNode(children, context)
  // for (let i = 0; i < children.length; i++) {
  //   const child = children[i]
  //   // 递归调用 genNode ->  switch 选中中间层 compound  -> 生成 + 号
  //   genNode(child, context)
  // }
  push(')')


  // 伪实现
  // push(`_createElementBlock("div")`)
}



// 处理 tag | props | children 的逻辑
function genNodeList(nodes, context) {
  const { push } = context
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i]
    // 判断 
    if (isString(node)) {
      push(node)
    } else {
      genNode(node, context)
    }

    // 添加 ,
    if (i < nodes.length - 1) {
      push(', ')
    }
  }
}



// 处理 tag | props | children 是否为 undefined 的情况
function genNullable(args: any) {
  return args.map(arg => arg || 'null')
}


// 处理 {{messag}}-> 生成 render 函数的返回内容
function genExpression(node, context) {
  const { push } = context
  push(`${node.content}`)
}


// 处理 interpolation -> 插值类型 生成 render 函数返回的内容
function genInterpolation(node, context) {
  const { push, helper } = context
  // console.log('node', node)
  // console.log('context', context)

  push(`${helper(TO_DISPLAY_STRING)}(`)
  // 解析插值中的 message -> 也是 Text , 递归调用 genText 解析
  genNode(node.content, context)
  push(")")

  // 伪实现 
  // push(`_toDisplayString(_ctx.message)`)
}



// 处理 text -> 生成 render 函数的返回内容
function genText(node: any, context: any) {
  const { push } = context
  push(`'${node.content}'`)
}


// 创建 基于 code 的全局上下文对象
function createCodegenContext() {
  const context = {
    // 存入 code
    code: '',
    // += 的行为 
    push(source) {
      context.code += source
    },
    helper(key) {
      return `_${helperMapName[key]}`
    }
  }

  return context
}

