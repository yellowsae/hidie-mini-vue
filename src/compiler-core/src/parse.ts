
// 初始化 compiler-core 的主函数流程 

import { NodeTypes } from "./ast"


const enum TagType {
  Start,
  End
}

// 导出 baseParse 函数 - 并且接收 content 模板的字符串 {{message}} 
export function baseParse(content: string) {

  // 1. 定义一个 context 全局上下文对象
  // 传入 "{{messages}}" 这个字符串 
  const context = createParseContext(content)

  // 2. 把根节点这个概念抽离为函数 
  // parseChildren 传入 context
  // 初始， 传入 ""
  // 改为 [] 模拟栈，把标签传入进去 
  return createRoot(parseChildren(context, []))
}



// 3. 继续抽离 children 
function parseChildren(context, ancestors) {  // 传入 parseTag 解析好的标签
  // 接收 context -> 后续处理 基于 context 生成 节点树
  // 功能： 返回一个数组 ，  [children]


  // 1. 定义一个 节点树 -> 也叫 抽象语法树
  const nodes: any = []


  // 3. 当解析完 text 后，后面还有内容: {{message}}</div>
  // 这里定义一个死循环，让它一直去解析后面的内容 
  // 当 source 没有值时候结束循环
  // 当 遇到 </div> 结束标签时候 停止循环
  // 使用 isEnd() 封装
  // 传入 parseTag 解析好的标签
  while (!isEnd(context, ancestors)) {
    let node;
    // 重构 - 简化 context.source
    let s = context.source
    // 增加 判断，当 context.source 是以 {{ 开头 时在进行插值的替换
    if (s.startsWith("{{")) {
      // 2. 调用 parseInterpolation() 返回出来的节点
      // 传入 context
      node = parseInterpolation(context)
    } else if (s[0] === '<') {
      // 判断 Element 并且解析 Element 的逻辑 

      // 判断第二个字符 是否是 a-z
      if (/[a-z]/i.test(s[1])) {
        // console.log("parse Element") // 命中 
        // 解析 Element 的逻辑
        // 把返回值放到 node 中
        node = parseElement(context, ancestors)
      }
    }


    // 默认 解析 Text , 当没有命中 Element 和 插值 时，解析 Text
    if (!node) { // 如果 node 没有值时，解析 Text
      node = parseText(context)
    }

    // 把生成的节点添加到 节点树 中 
    nodes.push(node)
  }

  // 最后返回 nodes -> 交给 createRoot 
  return nodes
}

// 判断是否循环的函数
function isEnd(context, ancestors) {
  // 2. 当 遇到 </div> 结束标签时候 停止循环
  let s = context.source

  // 使用 ancestors 栈中的数据 
  // 当解析到 结束标签时 
  if (s.startsWith("</")) {
    // 与 ancestors 中的数据进行对比 
    for (let i = 0; i < ancestors.length; i++) {
      const tag = ancestors[i].tag

      // 截取出 div  对比  tag  
      if (s.slice(2, 2 + tag.length) === tag) {
        return true
      }
    }
  }


  /** 
   * 因为这里是写死的标签 </div> 
   * 分析：这个div 是 解析 element.children 时 一开始的 <$>, 所以可以基于这个开始的标签 进行验证
   * 解决： 在  解析 element.children 传入解析好的 element.tag
   * 
  */
  //  改为 parseTag
  // 当 parseTag 为空时不用管
  // if (parseTag && s.startsWith(`</${parseTag}>`)) {
  //   // if (s.startsWith("</div>")) {
  //   // 返回 true
  //   return true
  // }
  // 1. 当 source 有值时候，返回 false
  return !context.source

}

// 解析 Text 的逻辑 
function parseText(context) {
  // 2. 增加 解析 Text 的判断 
  // 默认的解析 还是以 context.source.length 去获取 Text
  let endIndex = context.source.length
  // 如果 context.source 有 {{

  // 解决 context.source 中有嵌套标签的逻辑
  // 增加 < 
  let endTokens = ["<", "{{"]

  // 循环 判断
  for (let i = 0; i < endTokens.length; i++) {
    // 判断 context.source 中是否有 {{ 
    let index = context.source.indexOf(endTokens[i])
    // 如果 context.source 中同时有 < 和 {{,  尽可能往左取
    if (index !== -1 && endIndex > index) {
      // 如果 text 有 {{
      // 修改 endIndex 的值  
      endIndex = index
    }

  }



  // 1. 获取 content 内容 
  // 值就是 context.source
  // const content = context.source.slice(0, context.source.length);
  // 替换 index
  const content = context.source.slice(0, endIndex);
  // 此时解析出来的 content 为, hi, {{message}}</div>
  // 进行完 第二步 完， content 解析到 hi, 
  // console.log("-------", content);

  // 2. 删除 解析后的 text 
  advanceBy(context, content.length);
  // console.log(context.source)

  // 伪实现 
  return {
    // 类型
    type: NodeTypes.TEXT,
    // 解析出来的内容
    content,
  }
}


// 解析 Element 的逻辑
function parseElement(context: any, ancestors) {
  // console.log(context)   // 传入 { source: '<div></div>' }
  // 抽离解析 Element 的逻辑
  // 添加解析的类型
  const element: any = parseTag(context, TagType.Start)


  // 收集 tag 到 栈 ancestors 中
  ancestors.push(element)

  // 1. element 具有 children ，因为 element 可能是嵌套的
  // 所以 这里 递归调用 parseChildren 
  // 传入 解析好的 element.tag 
  element.children = parseChildren(context, ancestors)

  // 当解析好 tag 后，把栈中 对应的的数据 弹出来 
  ancestors.pop()



  // 当标签不匹配时，在 parseTag 出错 
  // 因为 开始标签 和 结束标签 不一致 
  // console.log("-----", element.tag, context.source) // ----- span </div>
  // 增加判断，如果匹配上标签 才执行 parseTag 
  if (context.source.slice(2, 2 + element.tag.length) === element.tag) {
    // 二次解析 Element 的逻辑 
    parseTag(context, TagType.End)
    // console.log(context.source)  // 删除全部标签代码 
  } else {
    // 如果标签不匹配 , 抛出错误 
    throw new Error(`缺少结束标签:${element.tag}`)
  }


  // 返回解析好的 Element 对象
  return element
}


// 抽离 解析 tag 的逻辑 
function parseTag(context, type: TagType) {
  // 1. 解析 div   tag 
  // 通过正则 匹配 div
  const match: any = /^<\/?([a-zA-Z][^\s/>]*)/.exec(context.source)

  // console.log(match)  // 1. <div  2. </div
  const tag = match[1]


  // 2. 删除处理完成的代码 
  advanceBy(context, match[0].length)

  advanceBy(context, 1)

  // 如果类型 === 结束标签 直接返回 
  if (type === TagType.End) return

  return {
    type: NodeTypes.ELEMENT,
    // 替换
    tag
  }
}


// 4，抽离解析 插值 {{}} 的内容 
function parseInterpolation(context) {


  // 因为这个的 message 是写死的，并且生成 节点
  // 目标：去除 {{ }} 拿到 message

  // 定义一个变量 为 {{  }}
  const openDelimiter = "{{"
  const closeDelimiter = "}}"
  // 1. 查找 }} 的索引 
  // indexOf("}}", 2) 查找 }}  从{{ 的位置开始查找
  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)

  // 2. 删除 {{  
  // 重构
  advanceBy(context, openDelimiter.length)
  // context.source = context.source.slice(openDelimiter.length)

  // 3. 通过 closeIndex 获取到 message 的长度 
  // -2 就是 减去 }} 的长度  
  const rawContextLength = closeIndex - openDelimiter.length // 得到 message 的长度 

  // 4. 获取 messages
  const rawContent = context.source.slice(0, rawContextLength)
  // 6 去除 " message " 的空格 空格 
  const content = rawContent.trim()


  // 5. 删除 }}
  // 重构
  advanceBy(context, rawContextLength + closeDelimiter.length)
  // context.source = context.source.slice(rawContextLength + closeDelimiter.length)



  return {
    // 解析的插值
    // 类型
    // 因为类型都是字符串的， 可以通过枚举重构 
    // type: "interpolation",
    type: NodeTypes.INTERPOLATION,
    // 内容
    content: {
      // message 的类型
      type: NodeTypes.STATEFUL_COMPONENT,
      // content 内容就是 message
      // content: "message"
      // 替换 动态的 message
      content: content
    }
  }
}


// 重构删除 字符串的逻辑
function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length)
}

// 2. 把根节点这个概念抽离为函数 
// 根节点对象 
function createRoot(children) {
  // 功能 : 返回一个对象， 这个对象具有一个 children
  return {
    children
  }
}

// 1. 定义一个 context 全局上下文对象
// 用于处理  "{{messages}}"
function createParseContext(content: string): any {
  // 功能 : 返回一个对象 
  // 这个对象具有一个 source ； 并把 content 传入 source ， 后续更具 source 操作
  return {
    source: content
  }
}
