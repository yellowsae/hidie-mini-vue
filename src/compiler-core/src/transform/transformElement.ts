

// 处理 添加 helper 函数的逻辑 

import { NodeTypes } from "../ast";
import { CREATE_ELEMENT_BLOCK } from "../runtimeHelpers";


export function transformElement(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    // 在这里加上 helper
    // 需要有一个 context


    // 返回一个函数， 也就是在 transform 需要收集的 退出函数 
    return () => {
      // 传入使用的常量  createElementBlock
      context.helper(CREATE_ELEMENT_BLOCK)


      // 中间处理层 



      // 1. tag 
      // tag 加上 ''  
      const vnodeTag = `'${node.tag}'`


      // 2. props 
      let vnodeProps;


      // 3. children

      const children = node.children
      // 中间层
      let vnodeChildren = children[0]



      // 4. 封装 element
      const vnodeElement = {
        type: NodeTypes.ELEMENT,
        tag: vnodeTag,
        props: vnodeProps,
        children: vnodeChildren
      }


      // 把 vnodeElement 赋值给 node ， 赋值到 codegenNode 上
      node.codegenNode = vnodeElement
    }
  }
}
