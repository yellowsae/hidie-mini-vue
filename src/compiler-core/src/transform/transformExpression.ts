

// 实现 _ctx 的引入使用 
// 在调用 transfrom() 时 传入给 option， 用于替换 _ctx

import { NodeTypes } from "../ast";


// 在  transfrom() 遍历递归 children 时 会调用  
export function transfromExpression(node) {


  // 1. 不会直接触发
  // 2. 触发时候 node 应该包在插值 里面 
  if (node.type === NodeTypes.INTERPOLATION) {
    // 获取 message -
    // const rawContent = node.content.content

    //  赋值 _ctx.message
    // node.content.content = "_ctx." + rawContent


    // 重构 
    node.content = processExpresstion(node.content)
  }
}
function processExpresstion(node: any) {
  node.content = `_ctx.${node.content}`
  return node
}

