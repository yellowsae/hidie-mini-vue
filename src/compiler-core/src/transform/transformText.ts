

// 实现中间层的逻辑 

import { NodeTypes } from "../ast";

export function transformText(node, context) {

  function isText(node) {
    return node.type === NodeTypes.TEXT || node.type === NodeTypes.INTERPOLATION
  }


  // 当是 element 类型时， 才会添加中间层
  if (node.type === NodeTypes.ELEMENT) {


    const { children } = node

    // 定义一个容器，用于收集 对应的数据
    let currentContainer


    for (let i = 0; i < children.length; i++) {
      // 拿到 child
      const child = children[i]

      // 1. 判断 它是 text 还是 插值 
      if (isText(child)) {
        // 2. 如果是 text , 才会查看它的 下一个(相邻节点) 是不是 插值 ， 如果是 创建中间层， 把它们收集起来 , 才能用 + 拼接

        for (let j = i + 1; j < children.length; j++) {
          // 获取到 child 的下一个节点 
          const next = children[j]

          // 再次判断 node
          if (isText(next)) {
            // 如果是, 进行收集 , 创建对应的容器

            // 初始化
            if (!currentContainer) {
              // 把 text 节点 替换为 中间层 compound
              // 中间层
              currentContainer = children[i] = {
                // 声明也需要和 节点一样 
                type: NodeTypes.COMPOUND_EXPRESSION,
                // 内容 存入 children 中
                children: [
                  child
                ]
              }
            }

            // push + 号 
            currentContainer.children.push(` + `)
            // 后续 只需要改变 children 内容
            currentContainer.children.push(next)

            // 删除 next
            children.splice(j, 1)
            j--

          } else {
            // 如果不是 text 或 插值
            // 停止收集 
            currentContainer = undefined
            break
          }
        }
      }
    }
  }
}
