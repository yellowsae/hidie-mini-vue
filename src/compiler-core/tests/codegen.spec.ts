import { generate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"
import { transformElement } from "../src/transform/transformElement"
import { transfromExpression } from "../src/transform/transformExpression"
import { transformText } from "../src/transform/transformText"


describe('codegen', () => {

  // 解析 string 
  it('string', () => {

    // hi 转为 render 返回的字符串 
    const ast = baseParse('hi')
    transform(ast)

    // 把 ast 树传给 generate() 
    const { code } = generate(ast)

    // 断言
    // 使用快照测试 -> 把 code 拍个照片， 然后进行对比
    // 1. 抓 bug 
    // 2.  更新 快照
    // 生成 snapshot 文件夹，存放 快照
    expect(code).toMatchSnapshot()
  })



  // 解析 插值 {{}}
  it('interpolation', () => {
    const ast = baseParse('{{ message }}')

    // 传入 需要生成 _ctx 的函数 
    transform(ast, {
      nodeTransformer: [transfromExpression]
    })
    // code 
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })


  // 实现 element
  it.skip('element', () => {
    const ast = baseParse('<div></div>')

    // 传入 需要生成 createElementBlock 的函数 
    transform(ast, {
      nodeTransformer: [transformElement]
    })
    // code 
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })


  // 实现 3 中联合类型 
  it('element', () => {
    const ast: any = baseParse('<div>hi, {{message}}</div>')

    // 传入 需要生成 createElementBlock 的函数 
    transform(ast, {
      // 换位 -> 最后才去执行 
      nodeTransformer: [transformText, transformElement]
    })

    // 检查看中间层 
    console.log('ast--------', ast, 'compound-------', ast.codegenNode.children)

    // code 
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

})
