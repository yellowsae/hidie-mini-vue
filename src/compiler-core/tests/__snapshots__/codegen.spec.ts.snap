// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"import { toDisplayString as _toDisplayString, createElementBlock as _createElementBlock } from \\"vue\\"
return function render(_ctx, _cache) {return _createElementBlock(\\"div\\", null, 'hi, ' + _toDisplayString(_ctx.message))}"
`;

exports[`codegen element 2`] = `
"import { createElementBlock as _createElementBlock, toDisplayString as _toDisplayString } from \\"vue\\"
return function render(_ctx, _cache) {return _createElementBlock(\\"div\\")}"
`;

exports[`codegen interpolation 1`] = `
"import { toDisplayString as _toDisplayString } from \\"vue\\"
return function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `
"
return function render(_ctx, _cache) {return 'hi'}"
`;
